onnxruntime/core/graph/graph_utils.cc

```diff
@@ -445,6 +465,65 @@ size_t RemoveNodeOutputEdges(Graph& graph, Node& node) {
   return output_edges.size();
 }
 
-}  // namespace graph_utils
+void ReplaceNodeOutput(Graph& graph, Node& node, Node& replacement, int replacement_output_idx) {
+  std::vector<GraphEdge> output_edges = GetNodeOutputEdges(node);
+
+  if (!output_edges.empty()) {
+    // all edges should have the same source index so double-check we don't do an invalid replacement
+    const auto output_idx = output_edges[0].src_arg_index;
 
+    const auto& replacement_name = replacement.MutableOutputDefs()[replacement_output_idx]->Name();
+
+    // Remove the output edges of the node first
+    RemoveGraphEdges(graph, output_edges);
```

Comment:
Here remove only the edge that is going to be replaced, not all output edges of the node (also update comment).

Reply:
Added variant of GetOutputEdges that restricted to the provided output index. 
Updated ReplaceNodeOutput to require the output index being replaced to be specified.